### PR TITLE
PipeScript implementation from hackathon

### PR DESCRIPTION
**Pull request by rodneykinney**
This is PipeScript, ready for real-world use.

@markschaake Thanks for taking a look.
@schmmd @OyvindTafjord FYI

### CODE REVIEW

File: src/main/scala/org/allenai/pipeline/Pipescript.scala [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+object PipeScript {
+  case class Package(id: String, source: URI) {
+    def scriptText: String = s"""package {id:"${id}", source:"${source}"}"""
+  }
+
+  /** A single line in a WorkflowScript that maps to a pipeline step
+    *
+    * Example StepCommand:
+    * {{{
+    * {in:$scripts/ExtractArrows.py} -i {in:./png, id:pngDir} -o {out:arrowDir, type:dir}
```
Comment (markschaake) [inline src/main/scala/org/allenai/pipeline/Pipescript.scala]: this example is out of date. Needs a `run` prefix

File: src/main/scala/org/allenai/pipeline/PipeScriptInterpreter.scala [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+import java.net.URI
+
+import org.allenai.pipeline.PipeScript._
+
+import scala.collection.mutable
+
+/** Interprets a PipeScript script and applies its workflow to a Pipeline
+  * @param pipeline
+  */
+class PipeScriptInterpreter(val pipeline: Pipeline) {
```
Comment (markschaake) [inline src/main/scala/org/allenai/pipeline/PipeScriptInterpreter.scala:13]: Nice rename :+1:

File: src/main/scala/org/allenai/pipeline/PipescriptSources.scala [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package org.allenai.pipeline
+
+import java.net.URI
+
+/** Created by rodneykinney on 8/20/15.
```
Comment (markschaake) [inline src/main/scala/org/allenai/pipeline/PipescriptSources.scala]: hmmm... I thought _I_ created this thing :)

File: src/universal/README.html [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    -outTest {output:testImages, type:dir}
+    
+run python {file:TrainModel.py, package: scripts}
+    -images {ref: trainImages}
+    -output {output:modelFile}
+    
+ run python {file: EvalModel.py, package: scripts}
+     -images {ref:testImages}
+     -model {ref:modelFile}
+     -output {output: evaluation, suffix:&quot;.txt&quot;}</code></pre>
```
Comment (markschaake) [inline src/universal/README.html:69]: why HTML and not Markdown?

Comment (rodneykinney) [inline src/universal/README.html:69]: This is an HTML version of the `PipeScript.md` file. I made it for the
binary release. If you do `sbt universal:packageBin` you get a zip file
with bin and lib directories and this README.

File: src/main/scala/org/allenai/pipeline/PipescriptSources.scala (hunk 3 above)
Comment (rodneykinney) [inline src/main/scala/org/allenai/pipeline/PipescriptSources.scala]: Heh. I should just take out all those author comments.

File: src/main/resources/logback-pipeScript.xml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<configuration>
```
Comment (rodneykinney) [inline src/main/resources/logback-pipeScript.xml:1]: This is used by the runPipeScript script in the binary distribution.

File: src/main/scala/org/allenai/pipeline/CodeInfo.scala [hunk 6]
```diff
@@ -1,11 +1,11 @@
 package org.allenai.pipeline
 
+import java.net.URI
+
 import com.typesafe.config.ConfigFactory
 
 import scala.collection.JavaConverters._
 
-import java.net.URI
```
Comment (rodneykinney) [inline src/main/scala/org/allenai/pipeline/CodeInfo.scala:7]: Organized the imports in bulk. Sorry for the extra noise.

File: src/main/scala/org/allenai/pipeline/ExternalProcess.scala [hunk 7]
```diff
@@ -1,314 +0,0 @@
...
-  *                        a String
-  *                        a Placeholder representing an input data file
-  *                        a Placeholder representing an output data file
-  *                        Examples:
-  *                        StringToken("cp") InputFileToken("src") OutputFileToken("target")
-  *                        StringToken("python") InputFileToken("script") StringToken("-o") OutputFileToken("output")
-  *
-  *                        Producers based on ExternalProcess should be created with class RunExternalProcess.
-  */
-class ExternalProcess(val commandTokens: CommandToken*) {
```
Comment (rodneykinney) [inline src/main/scala/org/allenai/pipeline/ExternalProcess.scala:25]: This was an earlier attempt at the `RunProcess` class

File: src/main/scala/org/allenai/pipeline/PipescriptParser.scala [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+    def stripQuotes(s: String) = s.substring(1, s.length - 1)
+
+    def unescape(s: String) = {
+      StringEscapeUtils.unescapeJava(s)
+    }
+  }
+
+  /** The parser combinator.  This is a subclass because parser combinators are not threadsafe. */
+  class Parser extends JavaTokenParsers {
```
Comment (rodneykinney) [inline src/main/scala/org/allenai/pipeline/PipescriptParser.scala:183]: Main difference is that commands can now be broken into multiple lines. End-of-line comments are also supported, but the comments are now discarded as whitespace, so we can't write them back to the portable script.

File: src/test/resources/pipeline/aristo-example/ablation-study.pipe [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  test2:"http://evaluation-framework.dev.allenai.org:8083/#/evaluation-sets/view/13b48240-1681-4911-908c-d3e7b00cf2dc/evaluation/e24bd1ed-20fb-4dea-966e-df42b8dde121"
+}
+
+# Packages
+package {id:aristoScripts, source:solver-combo/aristo-scripts}
+package {id:mathematicaScripts, source:solver-combo/mathematica-scripts}
+package {id:pythonScripts, source:solver-combo/python-scripts}
+
+# Get feature vectors for specific EF runs
+run bash {file:GetFeatureVectorsFromEF.sh, package:aristoScripts}
```
Comment (rodneykinney) [inline src/test/resources/pipeline/aristo-example/ablation-study.pipe:19]: Variables work in `run` commands. Very readable now!

**Comment by markschaake** [comment]
LGTM! Ship it! and let's get the new DAG UI over in S2 ASAP!

**Comment by OyvindTafjord** [comment]
Awesome, great job Rodney! The multiple lines look great, best argument for the "run" prefix so one can cleanly allow that. Maybe add a Pipescript link on the front page readme?

**Comment by schmmd** [comment]
Gasp.  Multiple lines!  I had thought that impossible...

More seriously, I like the braces `{}`.

**Pull request merged**